test/csr/csr.cc

```diff
@@ -0,0 +1,259 @@
+  if constexpr( RD0 ) {
+    assert( oldval == 0 );
+  } else {
+    assert( oldval == INIT );
+  }
+
+  // New value according to operation
+  switch( OP ) {
+  case CSROp::Write: assert( newval == ARG ); break;
+  case CSROp::Set: assert( newval == ( INIT | ARG ) ); break;
+  case CSROp::Clear: assert( newval == ( INIT & ~ARG ) ); break;
+  }
+}
+
+template<OpKind KIND, uint16_t CSR, size_t ARG, size_t INIT, bool RD0>
+void CSRTest2() {
+  CSRTest1<CSROp::Write, KIND, CSR, ARG, INIT, RD0>();
+  CSRTest1<CSROp::Set, KIND, CSR, ARG, INIT, RD0>();
+  CSRTest1<CSROp::Clear, KIND, CSR, ARG, INIT, RD0>();
+}
+
+template<OpKind KIND, uint16_t CSR, size_t INIT, bool RD0>
+void CSRTest3() {
+  CSRTest2<KIND, CSR, +size_t{ 0 }, INIT, RD0>();
+  CSRTest2<KIND, CSR, +size_t{ 1 }, INIT, RD0>();
+  CSRTest2<KIND, CSR, +size_t{ 2 }, INIT, RD0>();
+  CSRTest2<KIND, CSR, +size_t{ 3 }, INIT, RD0>();
+  CSRTest2<KIND, CSR, +size_t{ 4 }, INIT, RD0>();
+  CSRTest2<KIND, CSR, +size_t{ 8 }, INIT, RD0>();
+  CSRTest2<KIND, CSR, +size_t{ 16 }, INIT, RD0>();
+  CSRTest2<KIND, CSR, +size_t{ 31 }, INIT, RD0>();
+
+  if constexpr( KIND == OpKind::Reg ) {
+    CSRTest2<KIND, CSR, ~size_t{ 0 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, ~size_t{ 1 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, ~size_t{ 2 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, ~size_t{ 3 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, ~size_t{ 4 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, ~size_t{ 8 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, ~size_t{ 16 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, ~size_t{ 31 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, +size_t{ 256 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, ~size_t{ 256 }, INIT, RD0>();
+    CSRTest2<KIND, CSR, +size_t{ 1 } << sizeof( size_t ) * 8 - 1, INIT, RD0>();
+    CSRTest2<KIND, CSR, ~size_t{ 1 } << sizeof( size_t ) * 8 - 1, INIT, RD0>();
+  }
+}
+
+template<OpKind KIND, uint16_t CSR, bool RD0>
+void CSRTest4() {
+  CSRTest3<KIND, CSR, +size_t{ 0 }, RD0>();
+  CSRTest3<KIND, CSR, +size_t{ 1 }, RD0>();
+  CSRTest3<KIND, CSR, +size_t{ 2 }, RD0>();
+  CSRTest3<KIND, CSR, +size_t{ 3 }, RD0>();
+  CSRTest3<KIND, CSR, +size_t{ 4 }, RD0>();
+  CSRTest3<KIND, CSR, +size_t{ 8 }, RD0>();
+  CSRTest3<KIND, CSR, +size_t{ 16 }, RD0>();
+  CSRTest3<KIND, CSR, +size_t{ 31 }, RD0>();
+
+  if constexpr( KIND == OpKind::Reg ) {
+    CSRTest3<KIND, CSR, ~size_t{ 0 }, RD0>();
+    CSRTest3<KIND, CSR, ~size_t{ 1 }, RD0>();
+    CSRTest3<KIND, CSR, ~size_t{ 2 }, RD0>();
+    CSRTest3<KIND, CSR, ~size_t{ 3 }, RD0>();
+    CSRTest3<KIND, CSR, ~size_t{ 4 }, RD0>();
+    CSRTest3<KIND, CSR, ~size_t{ 8 }, RD0>();
+    CSRTest3<KIND, CSR, ~size_t{ 16 }, RD0>();
+    CSRTest3<KIND, CSR, ~size_t{ 31 }, RD0>();
+    CSRTest3<KIND, CSR, +size_t{ 256 }, RD0>();
+    CSRTest3<KIND, CSR, ~size_t{ 256 }, RD0>();
+    CSRTest3<KIND, CSR, +size_t{ 1 } << sizeof( size_t ) * 8 - 1, RD0>();
+    CSRTest3<KIND, CSR, ~size_t{ 1 } << sizeof( size_t ) * 8 - 1, RD0>();
+  }
+}
```

Comment:
When the operand is a register, we use values which do not fit in a zero-extended 5-bit immediate value. When the operand is an immediate, we can only use values `0` - `31`.